lib/vmpooler/metrics/promstats.rb

```diff
@@ -0,0 +1,379 @@
+# frozen_string_literal: true
+
+require 'prometheus/client'
+
+module Vmpooler
+  class Metrics
+    class Promstats < Metrics
+      attr_reader :prefix, :endpoint, :metrics_prefix
+
+      # Constants for Metric Types
+      M_COUNTER   = 1
+      M_GAUGE     = 2
+      M_SUMMARY   = 3
+      M_HISTOGRAM = 4
+
+      # Customised Bucket set to use for the Pooler clone times set to more appropriate intervals.
+      POOLER_TIME_BUCKETS = [1.0, 2.5, 5.0, 10.0, 20.0, 50.0, 100.0, 200.0, 500.0, 1000.0, 2000.0].freeze
```

Comment:
I'm concerned about how large the gaps get here. I think we are going to lose enough data that the metric is not very useful anymore. To retain useful data I would like to see something like buckets every 10 seconds up to 1800. I don't think that 1.0, 2.5, 5.0, or 10.0 is so meaningful compared to be able to preserve better granularity of the data.

I think it may be useful to have a few different types of buckets. Clone times should not often exceed 30, but in some cases I have seen reach as high as 900. The difference between 100 and 200 is pretty big, and I think it would be useful if the gaps were limited to, at largest, 30 seconds.

For time to ready state metrics many platforms are ready within 1-5 minutes, which would fall into 100, 200, or 500 in most cases. I think it would be very useful to be able to tell where within here we are falling. Again, I think anything more than 30 second gaps loses too much granularity. I would prefer to see consistent gaps, like every 30 seconds being the largest.

For platforms like Windows the worst end up being around 25 minutes when storage has higher latency. That means these would mostly fall into 1000 or 2000. However, I think that these gaps are much too big. For the larger end I do not think the gaps should get larger than one minute. Perhaps the upper end can be user configurable.

I think POOLER_TIME_BUCKETS should be broken up into at least two groups, one for short lived time buckets like clone times, migrations times, etc., and the other for longer operations, like time to ready state.

Reply:
I have thoughts here... working on a full response

Reply:
I like the sound of short-lived things having ~30 as the max and I can certainly see the benefits of consistent gaps. One thing we need to be careful of is too much cardinality though. I have pinged our Grafana reps for their thoughts on this subject too. 

When it comes to putting this data to use, I found https://grafana.com/docs/grafana/latest/getting-started/intro-histograms/ fairly helpful.